portal/models/qb_timeline.py

```diff
@@ -295,86 +360,118 @@ def ordered_qbs(user, classification=None):
-        else:
-            trace("single RP found")
-            next_rp = None
-
-        if sorted_rps:
-            raise ValueError("can't cope with > 2 RPs at this time")
-
-        if classification == 'indefinite':
-            current_qbds = indef_qbs_by_rp(current_rp.id, trigger_date=td)
-            next_rp_qbds = (
-                indef_qbs_by_rp(next_rp.id, trigger_date=td) if next_rp else
-                None)
-        else:
-            current_qbds = ordered_rp_qbs(current_rp.id, trigger_date=td)
-            next_rp_qbds = (
-                ordered_rp_qbs(next_rp.id, trigger_date=td) if next_rp else
-                None)
-
+        current_qbd, users_start, users_expiration = nextQBD(curRPD.qbds)
+        if not current_qbd:
+            trace("no current found in initial QBD lookup, bail")
+            return
+        if nextRPD:
+            next_qbd, next_start, next_expiration = nextQBD(nextRPD.qbds)
         while True:
-            # Advance both qb generators in sync while both RPs are defined
-            try:
-                current_qbd = next(current_qbds)
-            except StopIteration:
-                return
-            if next_rp_qbds:
-                try:
-                    next_rp_qbd = next(next_rp_qbds)
-                except StopIteration:
-                    return
-            users_start = calc_and_adjust_start(
-                user=user, qbd=current_qbd, initial_trigger=td)
-            users_expiration = calc_and_adjust_expired(
-                user=user, qbd=current_qbd, initial_trigger=td)
-
-            # if there's still a next and current gets retired before this
-            # QB's expiration, switch over *unless* user submitted a matching
-            # QNR on the current RP
-            if (next_rp_qbds and current_retired < users_expiration and
-                    not user_qnrs.earliest_result(
-                        qb_id=current_qbd.questionnaire_bank.id,
-                        iteration=current_qbd.iteration)):
-                current_qbds = next_rp_qbds
-                current_qbd, current_retired = next_rp_qbd, next_retired
-                next_rp_qbds, next_retired = None, None
-                users_start = calc_and_adjust_start(
-                    user=user, qbd=current_qbd, initial_trigger=td)
-
-            # done if user withdrew before the next start
+            if nextRPD and curRPD.retired < users_expiration:
+                # if there's a nextRP and curRP is retired before the
+                # user's QB for the curRP expires, we look to transition
+                # the user to the nextRP.
+                #
+                # if however, the user has already posted QNRs for the
+                # currentRP one of two things needs to happen:
+                #
+                #   1. if *any* of the QNRs are for instruments which
+                #      deterministically (i.e over RPs {v2,v3}, "epic23" vs
```

Comment:
looks like *deterministically* used twice...

Reply:
right, i can't seem to wordsmith this one.  i added that on @mcjustin's request that i clarify what deterministic meant in this context.
do you have a suggestion, or should we call it good enough?

Reply:
@ivan-c  Reworked as a footnote, better?